gcc: stop passing --with-ld when cross compiling

pcc opened the pull request:
Passing --with-ld causes problems when injecting a custom linker wrapper (similar to useMoldLinker) because it causes collect2 to check the --with-ld path (which is a hardcoded path to the standard linker wrapper) before $PATH, whereas with custom linker wrappers we want gcc to search $PATH.

I verified that -print-prog-name=ld still prints an appropriate value. For example, when cross compiling for x86_64 on an aarch64 machine, the following is printed by the util-linux configure script (which uses -print-prog-name=ld [1]):

checking for ld used by x86_64-unknown-linux-gnu-gcc... x86_64-unknown-linux-gnu-ld

We still pass --with-as because the gcc spec file relies on this to find the assembler. This is probably fine for now as it is less common to want to override the assembler than the linker.

[1] https://sources.debian.org/src/util-linux/2.41-4/m4/lib-ld.m4/?hl=59#L59


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [X] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

pcc commented:
This causes some build failures in rustc and python, will need to be fixed
```
       > error: linking with `/nix/store/5h9kdgskija8ywh9ns4ppsiw03nn0i37-riscv64-unknown-linux-gnu-gcc-wrapper-14.2.1.20250322/
bin/riscv64-unknown-linux-gnu-cc` failed: exit status: 1
       >   |
       >   = note:  "/nix/store/5h9kdgskija8ywh9ns4ppsiw03nn0i37-riscv64-unknown-linux-gnu-gcc-wrapper-14.2.1.20250322/bin/riscv
64-unknown-linux-gnu-cc" "-Wl,--version-script=/build/rustclcU4f4/list" "-Wl,--no-undefined-version" "/build/rustclcU4f4/symbols
.o" "<10 object files omitted>" "<sysroot>-std/riscv64gc-unknown-linux-gnu/release/deps/std-3b8173506a178add.c6pay6bdgmx5087mbyx
zkpnwd.rcgu.rmeta" "<1 object files omitted>" "-Wl,--as-needed" "-Wl,-Bstatic" "/build/rustc-1.86.0-src/build/aarch64-unknown-li
nux-gnu/stage1-std/riscv64gc-unknown-linux-gnu/release/deps/{libpanic_unwind-42e6d64f97121b91.rlib,libobject-8412a257614bdb77.rl
ib,libmemchr-8e0172225714eb43.rlib,libaddr2line-f185b2f1c49634c0.rlib,libgimli-c106403c3924c239.rlib,librustc_demangle-018dff3c6
59a372a.rlib,libstd_detect-c3cc750216d4df5a.rlib,libhashbrown-fcf29ec5b3d68b69.rlib,librustc_std_workspace_alloc-c25db315821fd41
3.rlib,libminiz_oxide-137b08f6932b52be.rlib,libadler2-e5352d63b9c66083.rlib,libunwind-7ab4b70283c65038.rlib,libcfg_if-b44f20e73e
0b9504.rlib,liblibc-85e8ec90b6379391.rlib,liballoc-0eb8e08e12474821.rlib,librustc_std_workspace_core-3ef30d61741dc754.rlib,libco
re-dd4c083535af2fd2.rlib,libcompiler_builtins-ea6041baa08380d4.rlib}.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread"
 "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "<sysroot>-std/riscv64gc-unknown-linux-gnu/release/build/comp
iler_builtins-5ee7c4f4b2317ff2/out" "-L" "<sysroot>/lib/rustlib/riscv64gc-unknown-linux-gnu/lib" "-o" "<sysroot>-std/riscv64gc-u
nknown-linux-gnu/release/deps/libstd-3b8173506a178add.so" "-shared" "-Wl,-soname=libstd-3b8173506a178add.so" "-Wl,-z,relro,-z,no
w" "-Wl,-O1" "-nodefaultlibs" "-Wl,-z,origin" "-Wl,-rpath,$ORIGIN/../lib"
       >   = note: some arguments are omitted. use `--verbose` to show all linker arguments
       >   = note: collect2: fatal error: cannot find 'ld'
       >           compilation terminated.
```
and
```
error: builder for '/nix/store/dv8fqzfm6y36sw8ii1liaqpg68nkhj8v-python3-riscv64-unknown-linux-gnu-3.12.9-env.drv' failed with ex
it code 1;
       last 3 log lines:
       > created 241 symlinks in user environment
       > collect2: fatal error: cannot find 'ld'
       > compilation terminated.
```

pcc commented:
> This causes some build failures in rustc and python, will need to be fixed

Fixed by adding binutils-wrapper to $PATH in cc-wrapper.

trofi commented on pkgs/development/compilers/gcc/common/configure-flags.nix:
```diff
@@ -69,7 +69,6 @@ let
       "--with-as=${
         if targetPackages.stdenv.cc.bintools.isLLVM then binutils else targetPackages.stdenv.cc.bintools
       }/bin/${targetPlatform.config}-as"
-      "--with-ld=${targetPackages.stdenv.cc.bintools}/bin/${targetPlatform.config}-ld"
```
`Another option would be to pass `"--with-ld=${targetPlatform.config}-ld"` and rely on `PATH` lookup.` (I hope it works as is).

pcc replied:
I don't think it would work. [Here](https://github.com/gcc-mirror/gcc/blob/1c5cf7cc82d3a2afa5d4fb90049456be408c2a7a/gcc/collect2.cc#L1056) and [here](https://github.com/gcc-mirror/gcc/blob/1c5cf7cc82d3a2afa5d4fb90049456be408c2a7a/gcc/gcc.cc#L3107) is the logic in GCC to check `DEFAULT_LINKER` (the `--with-ld` value). As you can see, it passes the value directly to `access` and doesn't search `$PATH`.

trofi replied:
Oh, yeah. Great find!

trofi commented:
And as it changes `pkgs/build-support/cc-wrapper/add-flags.sh` you'll need to target `staging` branch as it causes world rebuild.

pcc commented:
> And as it changes `pkgs/build-support/cc-wrapper/add-flags.sh` you'll need to target `staging` branch as it causes world rebuild.

Done

pcc committed:
gcc: stop passing --with-ld when cross compiling

Passing --with-ld causes problems when injecting a custom linker
wrapper (similar to useMoldLinker) because it causes collect2 to
check the --with-ld path (which is a hardcoded path to the standard
linker wrapper) before $PATH, whereas with custom linker wrappers we
want gcc to search $PATH.

Because binutils-wrapper is not guaranteed to be present in $PATH,
prepend it to $PATH in cc-wrapper.

I verified that -print-prog-name=ld still prints an appropriate
value. For example, when cross compiling for x86_64 on an aarch64
machine, the following is printed by the util-linux configure script
(which uses -print-prog-name=ld [1]):

checking for ld used by x86_64-unknown-linux-gnu-gcc... x86_64-unknown-linux-gnu-ld

We still pass --with-as because the gcc spec file relies on this
to find the assembler. This is probably fine for now as it is less
common to want to override the assembler than the linker.

[1] https://sources.debian.org/src/util-linux/2.41-4/m4/lib-ld.m4/?hl=59#L59

pcc commented:
I resolved the merge conflict on staging by effectively reverting 371a9fa6ebb510d5bf6006ea19b4b74cef14450e which looks like an alternative attempt to partially solve the problem but doesn't seem to work for overriding the linker search path (checked on Fedora by doing
```
cd /tmp; touch foo.c; strace -s9999 -f -o strace.log gcc -B/tmp/foo- foo.c
```
and didn't find `foo-ld` in strace.log).

alyssais commented:
@mildsunrise what do you think?

mildsunrise commented:
thanks for the ping. first of all, 100% agree that `--with-ld` needs to go. we want to use the bintools supplied to the cc-wrapper, not the one supplied to the gcc build. I would've done that in #395000 had I noticed.

now, it seems that I was foreshadowing when I said

> Because the original option could still be useful in other situations I'm not keeping into account (such as gcc) I suggest keeping both

because gcc does indeed have a very special quirk: `-B` entries are ignored if they contain a suffix AND the tool is being executed by collect2 instead of directly by the gcc process. this means the example given by @pcc (`gcc -B/tmp/foo- foo.c`) ignores the option -- but `gcc -B/tmp/ foo.c` does honor it, and so does `gcc -B/tmp/foo- -print-prog-name=ld`. it's hard to justify this behavior, I personally think it's a bug. clang certainly does not do that. none of it is documented.

<details><summary>More in depth explanation</summary>
In particular what happens is that `collect2` ignores `GCC_EXEC_PREFIX` or `-B` options in `COLLECT_GCC_OPTIONS` when searching for tools (ld, nm, strip). It uses only `COMPILER_PATH` and `PATH` for that. I guess their rationale is that collect2, despite being a gcc internal, is supposed to have the same interface as ld...

Now, gcc does add `-B` options and `GCC_EXEC_PREFIX` to `COMPILER_PATH` when invoking `collect2`... except that it only adds those entries that exist as directories. This is because `COMPILER_PATH` isn't supposed to have paths with suffixes, I guess.
</details>

rather than changing the whole approach to use `$PATH` instead of `-B`, we can restore the suffix-less `-B` option and that should work around the bug. that way, non-buggy compilers like clang can still benefit of the advantages of `-B` (like being able to specify suffixes, releasing a hidden `cc.targetPrefix == bintool.targetPrefix` requirement, and being more hermetic as it is not inherited)

pcc commented:
> rather than changing the whole approach to use `$PATH` instead of `-B`, we can restore the suffix-less `-B` option and that should work around the bug.

Do you mean that we would also pass `-B` without the target prefix like the argument that was being passed before #395000? That was basically the approach that I tried in an earlier version of this change (https://github.com/NixOS/nixpkgs/commit/e1cdb2c123db9c4a6d6b68dcec127fce73d85eed, note that this was on the master branch so I didn't have your change). Unfortunately, when I tried that, it didn't work for cross-compiling because gcc doesn't add the target prefix when searching for the linker in `-B` paths.

mildsunrise commented:
If I'm reading the source correctly, [it does add the target](https://github.com/gcc-mirror/gcc/blob/fceb6022798b587c9111d0241aaff72602dcd626/gcc/collect2.cc#L870-L872), but I'm seeing that it only does it (1) for cross-compile builds, (2) not for mold or lld, and (3) only when searching ld, not the other tools. Horrible.

In any case, the behavior is the same whether you pass the directory to `-B` or `PATH` or `COMPILER_PATH`, right?

pcc commented:
`full_ld_suffixes` is only used when [searching $PATH](https://github.com/gcc-mirror/gcc/blob/fceb6022798b587c9111d0241aaff72602dcd626/gcc/collect2.cc#L1110).

mildsunrise commented:
Ah, yes, you're very much right. `$PATH` it is, then...

trofi commented:
Bisect says ac2c00e10fe959190bfba685e9615527ef3c12f5 `gcc: stop passing --with-ld when cross compiling` broke `bear` test in `staging` as:

```
$ nix build --no-link -f. -L bear
...
bear> The following tests FAILED:
bear>     1 - bear::func_test (Failed)
```

pcc commented:
Verified fixed on x86_64-linux and aarch64-linux, sent #410699

pcc commented:
> Bisect says that this merge broke `gnat` build with:
> 
> ```
> error while loading shared libraries: libisl.so.19: cannot open shared object file
> ```
> 
> #413309

Fix: #419162
